Fixes for Intel in debug mode

andrew-platt opened the pull request:
**This PR is ready for merging, pending review**

**Feature or improvement description**
The Intel compiler (ifort (IFORT) 2021.4.0 20210910 from the OneAPI install on Ubuntu) throws runtime errors without these changes.

1. In _ServoDyn_, the `TRANSFER` function was throwing an error at runtime about the result of the `TRANSFER` not matching the size of the `C_CHAR` array on the LHS.  Adding the optional size seems to fix this (probably not necessary for `avcOUTNAME`, but added to eliminate confusion... or from my failure to note it wasn't necessary at the time). 
2. Passing the various unallocated `y%WriteOutput` arrays to the `WrOutputLine` resulted in the following error at runtime in debug:
```
forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable WRITEOUTPUT when it is not allocated

Image              PC                Routine            Line        Source
FAST.Farm          0000000007220FC2  Unknown               Unknown  Unknown
FAST.Farm          0000000000AF4F99  fast_subs_mp_writ        4948  FAST_Subs.f90
FAST.Farm          0000000000AD84BB  fast_subs_mp_fast        4016  FAST_Subs.f90
FAST.Farm          0000000000AD79B2  fast_subs_mp_fast        3934  FAST_Subs.f90
FAST.Farm          0000000000867D0F  fastwrapper_mp_fw         504  FASTWrapper.f90
FAST.Farm          00000000007ED72D  fast_farm_subs_mp        1711  FAST_Farm_Subs.f90
FAST.Farm          000000000089027F  MAIN__                    109  FAST_Farm.f90
FAST.Farm          0000000000406E12  Unknown               Unknown  Unknown
libc-2.31.so       00007FE06F5E00B3  __libc_start_main     Unknown  Unknown
```

**Related issue, if one exists**
Related conversation on PR #664:  https://github.com/OpenFAST/openfast/pull/664/files/2d5a2cd5fc456a77a6dc6cca444e3e0763d5548c..41579a7ed30e0f05c47ebd5260b47852eee85d8d#diff-bc64b5f80cb30c32b06ed3e52051ba117d0f3e765f375de9cd5bc7a81c6e0006

**Impacted areas of the software**
No changes in software operation for other compilers or compile settings is expected.

**Test results, if applicable**
No test results change with this.  And no, I'm not writing a unit test for this...

andrew-platt committed:
SrvD: types transfer size error with Intel debug

The runtime of Intel compiled in debug mode was complaining about 3 transfers.  Added the optional size to clear the issue

andrew-platt committed:
[bugfix] allocate y%WriteOutput of unused modules

The Intel compiler (ifort (IFORT) 2021.4.0 20210910 from the OneAPI install on Ubuntu) does not like passing these without them being allocated and throws a runtime error in debug mode.  There may be a better workaround for this issue that I am unaware of.

andrew-platt commented:
@bjonkman, do you mind weighing in on the allocate issue?  I think you've seen something similar before and likely have a better solution than what I put in.

bjonkman commented:
@andrew-platt , your solution should work fine. An alternative method would be to add the `allocatable` attribute on the variable definitions in the `WrOutputLine` routine. I used that method in HydroDyn when I was debugging in there a while back, but don't think I've actually pushed any of those changes to my OpenFAST fork.

Out of curiosity, have you checked what is in those two arrays after the `transfer` function in ServoDyn when the source character string is shorter than the output character array? I'm guessing they had some undefined values before this change. Are they now repeating the source? or blank? or ???

andrew-platt committed:
Change y%WriteOutput attribute to "ALLOCATABLE" instead of allocating size zero

andrew-platt commented:
@bjonkman, the transfer was keeping the `c_null_char` after all.  I made a mistake in my quick code for counting the number of null characters by stepping through the array from index 1 to `len(avcOUTNAME)` instead of `size(avcOUTNAME)`.  The transfer is populating the rest of the array with null characters, in this case 12584 null characters following 17 characters in the name for a total of 12601 charcters.

so the code
```
   CHARACTER(KIND=C_CHAR)                    :: avcOUTNAME(p%avcOUTNAME_LEN)         ! OUTNAME (in: Simulation RootName; out: string for logging channels Name:Units;)
   avcOUTNAME  = TRANSFER( TRIM(dll_data%RootName)//C_NULL_CHAR,   avcOUTNAME, p%avcOUTNAME_LEN )

do i=1,p%avcOUTNAME_LEN
   if(avcOUTNAME(i) == C_NULL_CHAR)   cnt=cnt+1
enddo
print*,'avcOUTNAME: ',p%avcOUTNAME_LEN,' Null chars: ',cnt
write(CU,*) avcOUTNAME(1:p%avcOUTNAME_LEN)
```

produces:
```
 avcOUTNAME:        12601  Null chars:        12584
 LESinflow.T1.SrvD


[output truncated]
```

andrew-platt committed:
[bugfix] add ALLOCATABLE attribute to ElastoDyn.f90::SetBladeParameters

Intel ifort in debug (2021.4.0 20210910) failed as follows with regression test:

```
Start 25: 5MW_Land_BD_DLL_WTurb

25: forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable INPBL when it is not allocated
25:
25: Image              PC                Routine            Line        Source
25: openfast           000000000685BB82  Unknown               Unknown  Unknown
25: openfast           00000000031E3771  elastodyn_mp_ed_s        2042  ElastoDyn.f90
25: openfast           00000000030D1A31  elastodyn_mp_ed_i         160  ElastoDyn.f90
```

andrew-platt committed:
[bugfix] add ALLOCATABLE attribute to FAST_Solver.f90::Create_FullOpt1_UVector

Intel ifort in debug (2021.4.0 20210910) failed as follows with regression test:

```
    Start 26: 5MW_OC4Jckt_ExtPtfm

26: Test timeout computed to be: 5400
26: forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable U_BD_ROOTMOTION when it is not allocated
26:
26: Image              PC                Routine            Line        Source
26: openfast           000000000685C872  Unknown               Unknown  Unknown
26: openfast           00000000005AF2DF  fast_solver_mp_fu        2346  FAST_Solver.f90
26: openfast           0000000000696295  fast_solver_mp_so        5242  FAST_Solver.f90
```

andrew-platt committed:
[bugfix] ALLOCATABLE attribute in Morison.f90

Intel ifort in debug (2021.4.0 20210910) failed as follows with regression test:

routine SetNodeMG
```
    Start 21: 5MW_ITIBarge_DLL_WTurb_WavesIrr

21: Test timeout computed to be: 5400
21: forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable MGDEPTHS when it is not allocated
21:
21: Image              PC                Routine            Line        Source
21: openfast           000000000685C782  Unknown               Unknown  Unknown
21: openfast           0000000004E50CAC  morison_mp_setmem        1468  Morison.f90
21: openfast           0000000004E730C6  morison_mp_setupm        1814  Morison.f90
```

routine SetExternalHydroCoefs
```
    Start 21: 5MW_ITIBarge_DLL_WTurb_WavesIrr

21: Test timeout computed to be: 5400
21: forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable COEFMEMBERS when it is not allocated
21:
21: Image              PC                Routine            Line        Source
21: openfast           000000000685C872  Unknown               Unknown  Unknown
21: openfast           0000000004E53C66  morison_mp_setmem        1484  Morison.f90
21: openfast           0000000004E732AF  morison_mp_setupm        1814  Morison.f90
```

routine WriteSummaryFile
```
    Start 21: 5MW_ITIBarge_DLL_WTurb_WavesIrr

21: Test timeout computed to be: 5400
21: forrtl: severe (408): fort: (8): Attempt to fetch from allocatable variable OUTPARAM when it is not allocated
21:
21: Image              PC                Routine            Line        Source
21: openfast           000000000685CDC2  Unknown               Unknown  Unknown
21: openfast           0000000004E84289  morison_mp_moriso        2183  Morison.f90
21: openfast           0000000004CDAF94  hydrodyn_mp_hydro        1372  HydroDyn.f90
```

andrew-platt committed:
[bugfix] TRANSFER size in map.f90

Intel ifort in debug (2021.4.0 20210910) failed as follows with regression test:

```
    Start 21: 5MW_ITIBarge_DLL_WTurb_WavesIrr

21: Test timeout computed to be: 5400
21: forrtl: severe (408): fort: (33): Shape mismatch: The extent of dimension 1 of array INITINP is 255 and the corresponding extent of array <RHS expression> is 79
21:
21: Image              PC                Routine            Line        Source
21: openfast           000000000685E972  Unknown               Unknown  Unknown
21: openfast           00000000061B791F  map_mp_map_set_in        1064  map.f90
21: openfast           00000000061B6E4E  map_mp_map_read_i        1047  map.f90
```

andrew-platt committed:
[bugfix] add ALLOCATABLE HydroDyn_Output.f90

Intel ifort in debug (2021.4.0 20210910) failed during some test (see previous commits as I'm feeling lazy about copying the details right now)
